### PR TITLE
[Downloader] Parametrization Storage with countries.txt file name and with downloading folder.

### PR DESCRIPTION
**Pull request by bykoianko**
Данный PR позволяет создавать несколько экземпляров Storage с разными countries.txt и скачивать карты при помощи них в разные директории.

Это нужно, поскольку до обновления, работая с большими mwm, нам необходимо перед началом обновления, скачать маленькую mwm около клиента.

@syershov @ygorshenin @vng @deathbaba посмотрите по возможности.

https://trello.com/c/KaGuXfOM/2255-storage-counries-txt

### CODE REVIEW

File: storage/storage.cpp [hunk 1]
```diff
@@ -83,8 +83,15 @@ bool HasCountryId(vector<TCountryId> const & sortedCountryIds, TCountryId const
   return binary_search(sortedCountryIds.begin(), sortedCountryIds.end(), countryId);
 }
 
-Storage::Storage() : m_downloader(new HttpMapFilesDownloader()), m_currentSlotId(0)
+Storage::Storage(string const & countiesTxt /* = COUNTRIES_FILE */, string const & folder /* = string() */)
```
Comment (ygorshenin) [inline storage/storage.cpp]: s/countriesTxt/pathToCountriesFile/g?

File: storage/storage_tests/storage_tests.cpp [hunk 2]
```diff
@@ -1069,4 +1070,56 @@ UNIT_TEST(StorageTest_IsPointCoveredByDownloadedMaps)
...
   }
 }
+
+UNIT_TEST(StorageTest_TwoInstance)
+{
+  Platform & platform = GetPlatform();
+  string const writableDir = platform.WritableDir();
+
+  string const testFolder1 = string("testfolder1");
+  Storage storage1(COUNTRIES_FILE, testFolder1);
```
Comment (ygorshenin) [inline storage/storage_tests/storage_tests.cpp]: Why not to create subclass of Storage (like, TestStorage), which will call Platform::SetWritableDirForTesting() ? After that you'll be able to get rid of this explicit storage specification?

Comment (ygorshenin) [inline storage/storage_tests/storage_tests.cpp]: Ah, forget it, bad idea.

File: platform/local_country_file.hpp [hunk 3]
```diff
@@ -84,7 +84,8 @@ class LocalCountryFile
   friend string DebugPrint(LocalCountryFile const &);
   friend void UnitTest_LocalCountryFile_DirectoryLookup();
   friend void FindAllLocalMapsAndCleanup(int64_t latestVersion,
-                                         vector<LocalCountryFile> & localFiles);
+                                         vector<LocalCountryFile> & localFiles,
+                                         string const & folder);
```
Comment (ygorshenin) [inline platform/local_country_file.hpp]: s/folder/dataDir/g ?

File: platform/local_country_file_utils.cpp [hunk 4]
```diff
@@ -205,11 +207,13 @@ void FindAllLocalMapsInDirectoryAndCleanup(string const & directory, int64_t ver
   }
 }
 
-void FindAllLocalMapsAndCleanup(int64_t latestVersion, vector<LocalCountryFile> & localFiles)
+void FindAllLocalMapsAndCleanup(int64_t latestVersion, vector<LocalCountryFile> & localFiles,
```
Comment (syershov) [inline platform/local_country_file_utils.cpp]: Please fix parameters order, latestVersion must be after localFiles same as other methods.

File: platform/local_country_file_utils.hpp [hunk 5]
```diff
@@ -42,7 +43,8 @@ void FindAllLocalMapsInDirectoryAndCleanup(string const & directory, int64_t ver
 //
 // Also, this method performs cleanup described in a comment for
 // CleanupMapsDirectory().
-void FindAllLocalMapsAndCleanup(int64_t latestVersion, vector<LocalCountryFile> & localFiles);
+void FindAllLocalMapsAndCleanup(int64_t latestVersion, vector<LocalCountryFile> & localFiles,
```
Comment (syershov) [inline platform/local_country_file_utils.hpp]: Please fix parameters order.

File: platform/local_country_file.hpp (hunk 3 above)
Comment (bykoianko) [inline platform/local_country_file.hpp]: Done

File: platform/local_country_file_utils.cpp (hunk 4 above)
Comment (bykoianko) [inline platform/local_country_file_utils.cpp]: done

File: platform/local_country_file_utils.hpp (hunk 5 above)
Comment (bykoianko) [inline platform/local_country_file_utils.hpp]: done

**Comment by bykoianko** [comment]
Учел все комментарии + использовал ScopedDir и удалил первые 2 коммита, которые релизовывали ScopedFileRemove.
@ygorshenin @syershov PTAL

**Comment by bykoianko** [comment]
@ygorshenin @syershov PTAL

File: platform/local_country_file_utils.cpp [hunk 6]
```diff
@@ -127,11 +128,12 @@ bool DirectoryHasIndexesOnly(string const & directory)
 }
 }  // namespace
 
-void DeleteDownloaderFilesForCountry(CountryFile const & countryFile, int64_t version)
+void DeleteDownloaderFilesForCountry(CountryFile const & countryFile,
+                                     int64_t version, string const & folder)
```
Comment (ygorshenin) [inline platform/local_country_file_utils.cpp]: s/folder/dataDir/g

**Comment by syershov** [comment]
LGTM

File: platform/local_country_file_utils.hpp [hunk 7]
```diff
@@ -18,7 +18,8 @@ namespace migrate
 }
 
 // Removes all files downloader creates during downloading of a country.
-void DeleteDownloaderFilesForCountry(CountryFile const & countryFile, int64_t version);
+void DeleteDownloaderFilesForCountry(CountryFile const & countryFile,
```
Comment (ygorshenin) [inline platform/local_country_file_utils.hpp]: Please, replace folder by dataDir too. Also, create a versions that do not take a string parameter and just calls these new functions with Platform::WritableDir(), from my point of view it's quite obfuscating that empty dataDir means Platform::WritableDir().

**Comment by bykoianko** [comment]
@ygorshenin PTAL

File: platform/local_country_file_utils.cpp [hunk 8]
```diff
@@ -305,11 +325,19 @@ shared_ptr<LocalCountryFile> PreparePlaceForCountryFiles(CountryFile const & cou
...
 {
+  return GetFileDownloadPath(countryFile, file, version, string());
+}
+
+string GetFileDownloadPath(CountryFile const & countryFile, MapOptions file,
+                           int64_t version, string const & dataDir)
+{
   Platform & platform = GetPlatform();
   string const readyFile = GetFileName(countryFile.GetName(), file, version) + READY_FILE_EXTENSION;
+  string const dir = dataDir.empty() ? platform.WritableDir()
```
Comment (ygorshenin) [inline platform/local_country_file_utils.cpp]: Could you please create a short inline function, like:
```c++
string GetFullDirPath(string const & dataDir) {
  return dataDir.empty() ? platform.WritableDir() : my::JoinFol...
}
```
And use it without code duplication?

File: platform/local_country_file_utils.hpp [hunk 9]
```diff
@@ -18,7 +18,11 @@ namespace migrate
 }
 
 // Removes all files downloader creates during downloading of a country.
+// Note. The the maps are deleted from writable dir/|dataDir|/|version| directory.
+// If |dataDir| is empty (or is not set) the function delete maps from writable dir/|dataDir|/.
```
Comment (ygorshenin) [inline platform/local_country_file_utils.hpp]: is empty -> is not empty

File: platform/local_country_file_utils.hpp [hunk 10]
```diff
@@ -42,7 +46,11 @@ void FindAllLocalMapsInDirectoryAndCleanup(string const & directory, int64_t ver
 //
 // Also, this method performs cleanup described in a comment for
 // CleanupMapsDirectory().
-void FindAllLocalMapsAndCleanup(int64_t latestVersion, vector<LocalCountryFile> & localFiles);
+// Note. The the maps are looked for writable dir/|dataDir|/|version| directory.
+// If |dataDir| is empty (or is not set) the function looks for maps in writable dir/|dataDir|/.
```
Comment (ygorshenin) [inline platform/local_country_file_utils.hpp]: If |dataDir| is empty (or is not set) the function looks for maps in writable dir/|version/.

Right?

File: platform/local_country_file_utils.hpp [hunk 11]
```diff
@@ -42,7 +46,11 @@ void FindAllLocalMapsInDirectoryAndCleanup(string const & directory, int64_t ver
...
+void FindAllLocalMapsAndCleanup(vector<LocalCountryFile> & localFiles, int64_t latestVersion);
```
Comment (ygorshenin) [inline platform/local_country_file_utils.hpp]: Why |localFiles| is the first parameter? All output parameters should be last, right?

File: storage/storage.hpp [hunk 12]
```diff
@@ -83,9 +83,14 @@ class Storage
   // country were successfully downloaded.
   TUpdate m_update;
 
+  // If |m_dataDir| is not empty Storage will create version directories and download maps into
```
Comment (ygorshenin) [inline storage/storage.hpp]: s/into/in/ ?

File: storage/storage.hpp [hunk 13]
```diff
@@ -83,9 +83,14 @@ class Storage
...
+  // platform::WritableDir/|m_dataDir|/. Not empty |m_dataDir| can be used only for
+  // downloading maps to a special place but not for continue working with them from this place.
```
Comment (ygorshenin) [inline storage/storage.hpp]: I'd like to paraphrase last sentence like:
```c++
Non-empty |m_dataDir| can be used to instantiate Storage in a test environment.
```

File: platform/local_country_file_utils.cpp (hunk 8 above)
Comment (bykoianko) [inline platform/local_country_file_utils.cpp]: Done

File: platform/local_country_file_utils.hpp (hunk 9 above)
Comment (bykoianko) [inline platform/local_country_file_utils.hpp]: // If |dataDir| is empty (or is not set) the function delete maps from writable dir.

Comment (bykoianko) [inline platform/local_country_file_utils.hpp]: Поправил на вариант выше в ряде похожих мест.

File: platform/local_country_file_utils.hpp (hunk 10 above)
Comment (bykoianko) [inline platform/local_country_file_utils.hpp]: If |dataDir| is empty (or is not set) the function looks for maps in writable dir

File: platform/local_country_file_utils.hpp (hunk 11 above)
Comment (ygorshenin) [inline platform/local_country_file_utils.hpp]: https://google.github.io/styleguide/cppguide.html#Function_Parameter_Ordering

Comment (bykoianko) [inline platform/local_country_file_utils.hpp]: Right. Fixed.

Comment (syershov) [inline platform/local_country_file_utils.hpp]: If we move output parameter to end of parameters list, then i would like version and dataDir to be the first and second parameters in every function in this scope.

File: storage/storage.hpp (hunk 12 above)
Comment (bykoianko) [inline storage/storage.hpp]: Are you sure about it? I guess the both variant is valid. 
I fixed to save time.

File: storage/storage.hpp (hunk 13 above)
Comment (bykoianko) [inline storage/storage.hpp]: We need it not only for test environment but for production.

**Comment by bykoianko** [comment]
@ygorshenin PTAL

**Comment by ygorshenin** [comment]
LGTM

File: platform/local_country_file_utils.hpp [hunk 14]
```diff
@@ -18,7 +18,11 @@ namespace migrate
 }
 
 // Removes all files downloader creates during downloading of a country.
-void DeleteDownloaderFilesForCountry(CountryFile const & countryFile, int64_t version);
+// Note. The the maps are deleted from writable dir/|dataDir|/|version| directory.
+// If |dataDir| is empty (or is not set) the function delete maps from writable dir.
```
Comment (ygorshenin) [inline platform/local_country_file_utils.hpp]: nit: s/delete/deletes/

Comment (bykoianko) [inline platform/local_country_file_utils.hpp]: fixed

**Pull request merged**